### DISCUSSION TITLE
Change output language

### DISCUSSION THREAD

**Discussion by nic0711** [danielmiessler/Fabric#162] [Q&A]
Hi!
Is it possible to change the output language dynamically?

E.g. fabric --output it|de|fr ...

Thanks and keep on rocking.

**Comment by xssdoctor** [comment]
I may add this soon, but one option you have is setting the language in the context.md file I.e. please respond in Spanish)

**Reply by nic0711** [reply]
Thank you! That would be fantastic!

**Reply by RoeslerM** [reply]
that is a nice one, thank you. 
1. create a file $HOME/.config/fabric/context.md with the text "please reply in french" 
2. call fabric with -c option like fabric -scp summarize
and one gets a French summary

**Reply by nic0711** [reply]
Doesn't work for me... maybe an issue with LS Stuidio or mistral-7b-instruct-v0.2.Q8_0.gguf?

**Reply by nic0711** [reply]
The -c option won't work with ollama either.

**Reply by xssdoctor** [reply]
I thought I fixed that a few days ago. Can you git pull then pipx upgrade fabric and see if it works?

**Comment by nic0711** [comment]
Hi!
Thank’s for your reply

Got this warnings on update:
⚠️ Found a space in the home path. We heavily discourage this, due to multiple incompatibilities. Please
    check our docs for more information on this, as well as some pointers on how to migrate to a
    different home path.
…
zsh: unknown sort specifier

****
context.md has one line of text: please reply in german
ollama model: mixtral:8x7b
****

Same result after update:
echo "Eine Idee zur Aussage: Die Welt ist schön." | fabric -scp write_essay
output is in english

###
Betreff: Re: [danielmiessler/fabric] Change output language (Discussion #162)

I thought I fixed that a few days ago. Can you git pull then pipx upgrade fabric and see if it works?

**Reply by xssdoctor** [reply]
Ahh. So -c and -o only works without -s. If you want to copy or do output, dont stream

**Reply by nic0711** [reply]
Ok... so the "marked as answer" is wrong...  ;)
... but streaming would be nice.

Thank you!

**Reply by nic0711** [reply]
Sorry, "echo "An idea that coding is like speaking with rules." | fabric -cp write_essay" don't work.
Output is english

**Reply by nic0711** [reply]
it's a bit strange... english input (clipboard) --> mixed output

```
User@MacBook-Pro ~ % pbpaste | fabric -scp summarize
 # EIN SATZ ZUSAMMENFASSUNG:
Coding is like speaking, both involving rule-based expression of ideas with less flexibility in coding.

## HAUPTPUNKTE:

1. Coding and speaking share similarities such as adhering to rules (syntax/grammar).
2. Both are ways of expressing ideas, structuring thoughts for others or computers.
3. Speaking allows for flexible expression, while coding requires precision.
4. Strict rules in coding help learn precise thinking, benefiting communication skills.
5. Understanding coding assists in understanding speaking rules and enhances communication.

## HERVORHEBUNGEN:

1. Learning to code improves communication by enforcing precise thinking through strict rules.
2. Mastering the syntax of programming languages sharpens one's command over language structures.
3. Speaking and coding both require creativity in expressing ideas within their respective constraints.
4. Developing a coder's mindset helps one become more aware of how they communicate their thoughts.
5. Familiarity with coding improves overall problem-solving abilities, complementing communication skills.
```

**Comment by RoeslerM** [comment]
I updated my fabric installation, and my default parameters are -scp and I still get output in german.Regarding the „space in the $PATH“, that is a specific result of German localization in OSX, where the directory for binary application includes a blank.Typos (C) by iPhoneAm 17.04.2024 um 20:19 schrieb xssdoctor ***@***.***>:﻿
Ahh. So -c and -o only works without -s. If you want to copy or do output, dont stream

**Reply by nic0711** [reply]
OK... so you get german output? With "stream"? Cool!
Local LLM? Any "special" configs?

**Comment by RoeslerM** [comment]
I had problems with POETRY. the „blank“ message comes from this.once I moved to PYENV as tool for managing virtual Python environments, problems disappeared.My problems all rooted in unclear or not well managed Python environments. F.e. I had CONDA on my system and that created a lot of mess.Typos (C) by iPhoneAm 17.04.2024 um 20:36 schrieb Wolf ***@***.***>:﻿
OK... so you get german output? With "stream"? Cool!
Local LLM? Any "special" configs?

**Reply by nic0711** [reply]
maybe conda is the root cause... Switching to pyenv is going to be real fun, not. ;) 
thank's

**Reply by RoeslerM** [reply]
conda definitively caused a lot of mess in my setup. PIP and similar tools pointed to an environment that then did not work with the fabric setup. With the help of a colleague I then first removed the conda installation (following the public available removal procedure) and then installed pyenv. Once I had that installed and setup properly, the fabric installation worked well.

regarding language setting: 
I have created a file $HOME/.config/fabric/context.md and it includes the text "please respond in German".
with that the flag -c triggers German output.
The command 
 % echo "elektrisches Motorrad" | fabric -scp create_threat_scenarios
results in 
## BEDROHUNGSSZENARIEN

- Diebstahl des Motorrads durch physische Überwindung der Sicherheitssysteme.
- Hackerangriff auf die Motorradelektronik zur unbefugten Nutzung oder Kontrolle.
- Manipulation der Software zur Veränderung der Leistungsdaten oder Umgehung von Sperren.
- Datendiebstahl persönlicher Informationen aus dem vernetzten Motorradsystem.
- Sabotage durch Eingriffe in die Motorsteuerung, was zu Unfällen führen kann.
- Vandalismus, der zu physischen Schäden am Motorrad führt.
- Batteriediebstahl oder -schädigung als spezifisches Ziel wegen des hohen Werts.
- GPS-Spoofing zur Fehlleitung oder zum Diebstahl des Motorrads.
- Phishing-Angriffe auf den Besitzer zur Erlangung von Zugangsdaten.
- Ausnutzung von Schwachstellen in der Bluetooth-Verbindung für unbefugten Zugriff.

## BEDROHUNGSMODELLANALYSE
[...]

I suggest you go step by step.
1) get fabric up and running. The command "fabric --help" should respond properly.
2) check if OpenAI API works with "fabric -sp summarize"
3) then look into the language/context thing.
4) check if the -c flag works with OpenAI API (thats is what I am using)
At this stage you have a working fabric with OpenAI model
5) enhance your setup for another AI mode
l

**Reply by RoeslerM** [reply]
since about a few weeks the language setting via the context.md file does not work any more.
Only work around I found is "add into the pattern" (which is not really userfriendly, I know).

**Reply by nic0711** [reply]
Hi @RoeslerM! It would be fantastic, if we could get this feature back. ;)

**Comment by szulaw** [comment]
Hi did someone figure out how to change output language dynamically? I also create a pattern for translation but then the order of mark down file is in mess.

**Comment by lobodecrin-studio** [comment]
I love Fabric, and initially, I made my custom patterns so that the responses would always be in Spanish. But I wanted to use all the patterns that already exist in the repository, not just the custom ones.

So I had this idea: I generated a custom model with Ollama's Modelfile, using any LLM like llama3.1 or any other, and forced in that Modelfile that all responses are always in Spanish, regardless of whether the prompts from the patterns or the inputs are written in English or Spanish.

Now I'm using my custom ollama model in fabric and it works. I hope this helps.
